Update Trusted Advisor Service Limit Checks

arturoera opened the pull request:
According to AWS support, the checkid `eW7HH0l7J9` does not work, and will be deprecated anytime soon. This PR fixes this problem by adding the entire list of new checkids from: 
https://aws.amazon.com/premiumsupport/ta-iam/#table2

It also removes the CLI option to provide a region, and instead returns metrics with labels for each region, ex.:
`aws_vpc_used_total{region="us-west-1",resource="Internet gateways"} 1
aws_vpc_used_total{region="us-west-1",resource="VPC Elastic IP addresses (EIPs)"} 0
aws_vpc_used_total{region="us-west-1",resource="VPCs"} 1
aws_vpc_used_total{region="us-west-2",resource="Internet gateways"} 4
aws_vpc_used_total{region="us-west-2",resource="VPC Elastic IP addresses (EIPs)"} 21
aws_vpc_used_total{region="us-west-2",resource="VPCs"} 5`

Finally it updates the project to go modules.

arturoera committed:
moved to go modules

arturoera committed:
initial linting, inclusion for all regions metrics

arturoera committed:
moved structs to types.go

arturoera committed:
removed region from cli flag

arturoera committed:
added all list of service limits from aws

arturoera committed:
updated makefile

danielfm commented on core/support.go:
```diff
@@ -55,29 +142,38 @@ func NewSupportClient() *SupportClientImpl {
-		waitMs := *output.Status.MillisUntilNextRefreshable
-
-		glog.Infof("Refresh status is '%s', waiting %dms until the next refresh...", aws.StringValue(output.Status.Status), waitMs)
+		glog.Infof("Refresh status is '%s', waiting %d minutes until the next refresh...", refreshStatus, waitMs/60000)
```
We are reporting only the status of the last check, which seems misleading.

Since we are already reporting the error for a specific check inside the second `for`, maybe this log line is not providing any useful information and thus can be removed?

```suggestion
```

arturoera replied:
I think its still useful to have the refresh status output, I'm not sure about the duration, I think it does not hurt to leave it there so anyone looking at the program logs can easily see the program is just waiting for its next iteration. 
This iis how it looks:
`I1120 12:19:41.224842   32617 support.go:168] Refresh status is 'enqueued', waiting 59 minutes until the next refresh...`

danielfm replied:
Makes sense, although I'd remove the refresh status from the message, as it's just the refresh status for the last check and might not reflect the status of all checks.

arturoera replied:
agree, I added a commit to remove the refresh status., and only logs we are waiting 60 min for the next iteration:
`I1121 09:03:23.834109     536 support.go:156] Refreshing Trusted Advisor check 'kM7QQ0l7J9'...
I1121 09:03:23.990282     536 support.go:156] Refreshing Trusted Advisor check 'jL7PP0l7J9'...
I1121 09:03:24.099742     536 support.go:163] Waiting 60 minutes until the next refresh...`

danielfm commented on core/support.go:
```diff
@@ -55,29 +142,38 @@ func NewSupportClient() *SupportClientImpl {
+				continue
+			}
+
+			waitMs = *output.Status.MillisUntilNextRefreshable
```
Here, we overwrite `waitMs` according to the current check's `MillisUntilNextRefreshable`, and use the last one outside the loop to sleep.

I don't know if there are different refresh periods for different checks. If there are, maybe we can change the logic to keep the minimum `MillisUntilNextRefreshable` across all checks, and sleep for that period. Not very efficient (since we would try to refresh some checks that haven't reached the next update window) but simple to implement.

Another (more complex, even though more accurate) option would be to create a Goroutine for each check, and sleep for the correct amount of time in each check.

My suggestion is to go with the first option.

arturoera replied:
I doubled check each `MillisUntilNextRefreshable` for each check, all of them are closed to 1 hour, so I'll set it to that instead. 

Ex output (won't add this log to the logs):
`I1120 12:19:40.273771   32617 support.go:165]  Trusted Advisor check 'P1jhKWEmLa'... duration: '3599992'
I1120 12:19:40.364034   32617 support.go:165]  Trusted Advisor check 'dx3xfcdfMr'... duration: '3599993'
I1120 12:19:40.449686   32617 support.go:165]  Trusted Advisor check 'ru4xfcdfMr'... duration: '3599992'
I1120 12:19:40.685881   32617 support.go:165]  Trusted Advisor check 'ty3xfcdfMr'... duration: '3599992'
I1120 12:19:40.766649   32617 support.go:165]  Trusted Advisor check 'dx3xfbjfMr'... duration: '3599993'
I1120 12:19:40.856893   32617 support.go:165]  Trusted Advisor check 'dx8afcdfMr'... duration: '3599992'
I1120 12:19:40.951466   32617 support.go:165]  Trusted Advisor check 'hJ7NN0l7J9'... duration: '3599993'
I1120 12:19:41.038617   32617 support.go:165]  Trusted Advisor check 'lN7RR0l7J9'... duration: '3599992'
I1120 12:19:41.138710   32617 support.go:165]  Trusted Advisor check 'kM7QQ0l7J9'... duration: '3599991'
I1120 12:19:41.224817   32617 support.go:165]  Trusted Advisor check 'jL7PP0l7J9'... duration: '3599993'`

danielfm commented on core/support.go:
```diff
@@ -88,78 +184,76 @@ func (client *SupportClientImpl) DescribeServiceLimitsCheckResult() (*support.Tr
-			e.metricsLimit[resourceId] = newServerMetric(e.region, serviceNameLower, "limit_total", "Current limit of the given resource.", []string{"resource"})
+			serviceName := aws.StringValue(resource.Metadata[1])
+			serviceNameLower := strings.ToLower(serviceName)
+			glog.Infof("Updating Data,     Region: '%s', Service Name: '%s'", aws.StringValue(resource.Metadata[0]), serviceName)
```
Is this log really necessary?

arturoera replied:
not really, removed.

danielfm commented on go.mod:
```diff
@@ -0,0 +1,11 @@
+	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
+	github.com/modern-go/reflect2 v1.0.1 // indirect
+	github.com/prometheus/client_golang v1.2.1
+)
```
Is the current `Makefile` working after moving to go modules? Should we change the `Makefile` in any way to improve the development workflow now?

arturoera replied:
forgot to include the new Makefile in the PR, updating as well.

danielfm reviewed (changes requested):
Thanks for your contribution! Just a few comments and suggestions.

arturoera committed:
addressed PR questions

arturoera committed:
include makefile

arturoera committed:
dockerfile update

arturoera committed:
updated readme

arturoera commented:
Addressed all the comments, also updated the Makefile, Dockerfile, and readme

We have a running copy of this image in our 3 environments, it has worked pretty good so far, detecting aws limits about to be reached, and we also hooked it into Alertmanager and have good alerts for all the regions. Thank you for your work on this.

danielfm commented on Makefile:
```diff
@@ -1,21 +1,44 @@
-push: image
-	docker push $(IMAGE):$(TAG)
+docker-build:
+	docker build -t ${DOCKER_REPO}$(DOCKER_IMAGE):latest .
```
Does this `${DOCKER_REPO}` interpolate correctly, or should we change it to `$(DOCKER_REPO)`? The same applies to the commands in `docker-deploy` target.

arturoera replied:
it worked, however I changed the commands to add the `/`, so the repo name can be let alone. Ex: 
`make docker-deploy`
`docker tag danielfm/aws-limits-exporter:latest danielfm/aws-limits-exporter:0.2.0`
`docker push danielfm/aws-limits-exporter:0.2.0`

arturoera committed:
removes refreshstatus log

arturoera committed:
improve docker repo name

danielfm commented on Makefile:
```diff
@@ -1,21 +1,44 @@
+run-server: build
+	$(GOBIN)/aws-limits-exporter
+
+run-linux: build
+	$(GOBIN)/aws-limits-exporter
+
```
This target is duplicated, we can remove it.

danielfm reviewed (approved):
Thanks a lot for your contribution! :100:
